include/tracing_impl/nvtx.h

```diff
@@ -0,0 +1,190 @@
+ */
+
+#ifndef NVTX_H
+#define NVTX_H
+
+#if HAVE_NVTX_TRACING
+#include "nvToolsExt.h"
+
+static inline void nvtx_mark_domain(nvtxDomainHandle_t domain, const char* name, uint32_t color)
+{
+	const nvtxEventAttributes_t eventAttrib = {
+		.version = NVTX_VERSION,
+		.size = NVTX_EVENT_ATTRIB_STRUCT_SIZE,
+		.colorType = NVTX_COLOR_ARGB,
+		.color = color,
+		.messageType = NVTX_MESSAGE_TYPE_ASCII,
+		.message = { .ascii = name },
+	};
+	nvtxDomainMarkEx(domain, &eventAttrib);
+}
+
+static inline nvtxRangeId_t nvtx_start_domain(bool have_domain, nvtxDomainHandle_t domain, const char* name, uint32_t color) {
+	const nvtxEventAttributes_t eventAttrib = {
+		.version = NVTX_VERSION,
+		.size = NVTX_EVENT_ATTRIB_STRUCT_SIZE,
+		.colorType = NVTX_COLOR_ARGB,
+		.color = color,
+		.messageType = NVTX_MESSAGE_TYPE_ASCII,
+		.message = { .ascii = name },
+	};
+	if (have_domain)
+		return nvtxDomainRangeStartEx(domain, &eventAttrib);
+	else
+		return nvtxRangeStartEx(&eventAttrib);
+}
+
+static inline nvtxRangeId_t nvtx_start(const char* name, uint32_t color) {
+	return nvtx_start_domain(false, 0, name, color);
+}
+
+static inline void nvtx_end_domain(nvtxDomainHandle_t domain, nvtxRangeId_t id) {
+	nvtxDomainRangeEnd(domain, id);
+}
+
+static inline void nvtx_end(nvtxRangeId_t id) {
+	nvtxRangeEnd(id);
+}
+
+#define NCCL_OFI_TRACE_SEND_NVTX(dev, size, comm, msg_seq_num, request, nccl_req) do { \
+	if (NCCL_OFI_NVTX_TRACE_PER_COMM) { \
+		nvtxDomainHandle_t handle = ((nccl_net_ofi_rdma_send_comm_t*)comm) \
+			->nvtx_domain[msg_seq_num % N_NVTX_DOMAIN_PER_COMM]; \
+		get_send_data(request)->trace_id = nvtx_start_domain(true, handle, "Send", 0xeb9234); \
+	} \
+} while (0)
+
+#define NCCL_OFI_TRACE_SEND_END_NVTX(request) do { \
+	if (NCCL_OFI_NVTX_TRACE_PER_COMM) { \
+		nvtxDomainHandle_t handle = ((nccl_net_ofi_rdma_send_comm_t*)(request->comm)) \
+			->nvtx_domain[request->msg_seq_num % N_NVTX_DOMAIN_PER_COMM]; \
+		nvtx_end_domain(handle, get_send_data(request)->trace_id); \
+	} \
+} while(0)
+
+#define NCCL_OFI_TRACE_SEND_CTRL_RECV_NVTX(dev, rail_id, comm, msg_seq_num) do { \
+	nvtxDomainHandle_t handle; \
+	if (NCCL_OFI_NVTX_TRACE_PER_COMM) { \
+		handle = ((nccl_net_ofi_rdma_send_comm_t*)comm)->nvtx_domain[msg_seq_num % N_NVTX_DOMAIN_PER_COMM]; \
+		nvtx_mark_domain(handle, "Send_ctrl_recv", 0x00ffff); \
+	} \
+	if (NCCL_OFI_NVTX_TRACE_PER_DEV) { \
+		handle = ((nccl_net_ofi_rdma_device_t*)(comm->base.base.ep->device))->nvtx_domain[rail_id]; \
+		nvtx_mark_domain(handle, "Send_ctrl_recv", 0x00ffff); \
+	} \
+} while (0)
+
+#define NCCL_OFI_TRACE_SEND_CTRL_START_NVTX(dev, rail_id, comm, req, msg_seq_num) do { \
+	nvtxDomainHandle_t handle; \
+	if (NCCL_OFI_NVTX_TRACE_PER_COMM) { \
+		handle = ((nccl_net_ofi_rdma_recv_comm_t *)comm)->nvtx_domain[msg_seq_num % N_NVTX_DOMAIN_PER_COMM]; \
+		get_send_ctrl_data(req)->trace_id = nvtx_start_domain(true, handle, "Send_ctrl_start", 0x00ffff); \
+	} \
+	if (NCCL_OFI_NVTX_TRACE_PER_DEV) { \
```

Comment:
else if? or a static assert that NCCL_OFI_NVTX_TRACE_PER_DEV and NCCL_OFI_NVTX_TRACE_PER_COMM are mutually exclusive